File: ios/Pchook/Shared/Book.swift
import Foundation

struct APIResponse<T: Decodable & Sendable>: Decodable, Sendable {
    let status: Int
    let data: T
}

// MARK: - Enums

enum BookLanguage: String, CaseIterable, Identifiable {
    case fr = "FR"
    case en = "EN"

    var id: String { rawValue }

    var label: String {
        switch self {
        case .fr: "Français"
        case .en: "English"
        }
    }

    init?(apiValue: String?) {
        guard let value = apiValue?.trimmingCharacters(in: .whitespaces).uppercased() else { return nil }
        self.init(rawValue: value)
    }
}

enum BookFormatOption: String, CaseIterable, Identifiable {
    case pocket
    case paperback
    case hardcover
    case audiobook

    var id: String { rawValue }

    var label: String {
        switch self {
        case .pocket: "Poche"
        case .paperback: "Broché"
        case .hardcover: "Relié"
        case .audiobook: "Livre audio"
        }
    }

    init?(apiValue: String?) {
        guard let value = apiValue?.trimmingCharacters(in: .whitespaces).lowercased() else { return nil }
        self.init(rawValue: value)
    }
}

// MARK: - Requests

struct UpdateBookRequest: Encodable, Sendable {
    var title: String?
    var authors: [String]?
    var publisher: String?
    var publishedDate: String?
    var pageCount: Int?
    var genre: String?
    var synopsis: String?
    var isbn: String?
    var language: String?
    var format: String?
    var translator: String?
    var estimatedPrice: Double?
    var duration: String?
    var narrators: [String]?
    var personalNotes: String?
    var status: String?
    var readDate: String?
    var series: String?
    var seriesLabel: String?
    var seriesNumber: Int?
}

struct AnalyzeBookRequest: Encodable, Sendable {
    let imageBase64: String
    var ocrText: String?
}

struct ConfirmBookRequest: Encodable, Sendable {
    let previewId: String
    let status: String
    var overrides: ConfirmBookOverrides?
    var replaceBookId: String?
}

struct ConfirmBookOverrides: Encodable, Sendable {
    var title: String?
    var authors: [String]?
    var publisher: String?
    var pageCount: Int?
    var genre: String?
    var synopsis: String?
    var language: String?
    var format: String?
    var translator: String?
    var estimatedPrice: Double?
    var series: String?
    var seriesLabel: String?
    var seriesNumber: Int?
}

struct CreateReviewRequest: Encodable, Sendable {
    let rating: Int
    var readDate: String?
    var reviewNotes: String?
}

File: server/api-schema/book.ts
import { z } from 'zod'

export const awardSchema = z.object({
  name: z.string(),
  year: z.number().int().optional(),
})

export const publicRatingSchema = z.object({
  source: z.string(),
  score: z.number(),
  maxScore: z.number(),
  voterCount: z.number().int(),
  url: z.string().optional(),
})

export const bookSchema = z.object({
  id: z.string(),
  title: z.string(),
  authors: z.array(z.string()),
  publisher: z.string().optional(),
  publishedDate: z.string().datetime().optional(),
  pageCount: z.number().int().optional(),
  genre: z.string().optional(),
  synopsis: z.string().optional(),
  isbn: z.string().optional(),
  language: z.string().optional(),
  format: z.string().optional(),
  translator: z.string().optional(),
  estimatedPrice: z.number().optional(),
  duration: z.string().optional(),
  narrators: z.array(z.string()),
  personalNotes: z.string().optional(),
  status: z.string(),
  readDate: z.string().datetime().optional(),
  awards: z.array(awardSchema),
  publicRatings: z.array(publicRatingSchema),
  importSource: z.string().optional(),
  externalUrl: z.string().optional(),
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
})

export const bookPreviewSchema = z.object({
  previewId: z.string(),
  title: z.string(),
  authors: z.array(z.string()),
  publisher: z.string().optional(),
  publishedDate: z.string().optional(),
  pageCount: z.number().int().optional(),
  genre: z.string().optional(),
  synopsis: z.string().optional(),
  isbn: z.string().optional(),
  language: z.string().optional(),
  format: z.string().optional(),
  series: z.string().optional(),
  seriesLabel: z.string().optional(),
  seriesNumber: z.number().int().optional(),
  translator: z.string().optional(),
  estimatedPrice: z.number().optional(),
  duration: z.string().optional(),
  narrators: z.array(z.string()).optional(),
  awards: z.array(awardSchema),
  publicRatings: z.array(publicRatingSchema),
  coverImageBase64: z.string().optional(),
})
